src/Microsoft.SqlTools.ServiceLayer/LanguageServices/InteractionMetrics.cs

```diff
@@ -0,0 +1,97 @@
+﻿//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.SqlTools.ServiceLayer.Utility;
+
+namespace Microsoft.SqlTools.ServiceLayer
+{
+    /// <summary>
+    /// A class to calculate the value for the metrics using the given bucket
+    /// </summary>
+    public class InteractionMetrics<T>
+    {
+        /// <summary>
+        /// Creates new instance given a bucket of metrics
+        /// </summary>
+        public InteractionMetrics(int[] metrics)
+        {
+            Validate.IsNotNull("metrics", metrics);
+            if(metrics.Length == 0)
+            {
+                throw new ArgumentOutOfRangeException("metrics");
+            }
+
+            Counters = new ConcurrentDictionary<string, T>();
+            Metrics = IsSorted(metrics) ? metrics : metrics.OrderBy( i => i).ToArray();
```

Comment:
>metrics.OrderBy( i => i).ToArray( [](start = 52, length = 33)

If you do an in-place sort it looks like you could avoid an unnecessary additional array allocation..e.g., [https://msdn.microsoft.com/en-us/library/kwx6zbd4(v=vs.110).aspx](https://msdn.microsoft.com/en-us/library/kwx6zbd4(v=vs.110).aspx)